vpr/src/draw/search_bar.cpp

```diff
@@ -339,34 +420,135 @@ void warning_dialog_box(const char* message) {
+}
+
+/**
+ * @brief Creates a GdkEvent that simulates user pressing key "key"
+ * 
+ * @param key character value
+ * @param window GdkWindow
+ * @return GdkEvent Keypress event
+ */
+GdkEvent simulate_keypress(char key, GdkWindow* window) {
+    int charVal = (int)key;
+    //Creating event and adding properties
+    GdkEvent new_event;
+    new_event.key.type = GDK_KEY_PRESS;
+    new_event.key.window = window;
+    new_event.key.send_event = TRUE;
+    new_event.key.time = GDK_CURRENT_TIME;
+    new_event.key.keyval = gdk_unicode_to_keyval(charVal);
+    new_event.key.state = GDK_KEY_PRESS_MASK;
+    new_event.key.length = 0;
+    new_event.key.string = 0;
+    new_event.key.hardware_keycode = 0;
+    new_event.key.group = 0;
+    return new_event;
+}
+
+/**
+ * @brief Turns on autocomplete
+ * 
+ * This function enables the auto-complete fuctionality for the search bar. 
+ * Normally, this is pretty simple, but the idea is to have auto-complete appear as soon
+ * as the user hits the "Enter" key. To accomplish this, a fake Gdk event is created
+ * to simulate the user hitting a key.
+ * 
+ * @param app ezgl app
+ */
+void enable_autocomplete(ezgl::application* app) {
+    GtkEntryCompletion* completion = GTK_ENTRY_COMPLETION(app->get_object("Completion"));
+    GtkEntry* searchBar = GTK_ENTRY(app->get_object("TextInput"));
+
+    std::string searchType = get_search_type(app);
+    //Checking to make sure that we are on a mode that uses auto-complete
+    if (gtk_entry_completion_get_model(completion) == NULL) {
+        std::cout << "NO MODEL SELECTED" << std::endl;
+        return;
     }
+
+    //Getting input text
+    std::string oldText(gtk_entry_get_text(searchBar));
+
+    //Turning on completion
+    gtk_entry_set_completion(searchBar, completion);
+    gtk_entry_completion_complete(completion);
+
+    //Setting min key length to either 0 or 1 less than key length (max option)
+    gtk_entry_completion_set_minimum_key_length(completion, std::max(0, (int)(oldText.length() - 1)));
+
+    //If string len is 0, reutrning
+    if (oldText.length() == 0) return;
+
+    gtk_widget_grab_focus(GTK_WIDGET(searchBar));
+    std::string newText = (oldText.length() > 1) ? oldText.substr(0, oldText.length() - 1) : "";
+    gtk_entry_set_text(searchBar, newText.c_str());
+
+    //Creating a false event to insert the last character into the string
+    auto window = gtk_widget_get_parent_window(GTK_WIDGET(searchBar));
+    GdkEvent new_event = simulate_keypress(oldText.back(), window);
+    gdk_event_put(&new_event);
+}
+
+std::string get_search_type(ezgl::application* app) {
+    GObject* combo_box = (GObject*)app->get_object("SearchType");
+    gchar* type = gtk_combo_box_text_get_active_text(GTK_COMBO_BOX_TEXT(combo_box));
+    //Checking that a type is selected
+    if (type && type[0] == '\0') {
+        warning_dialog_box("Please select a search type");
+        app->refresh_drawing();
+        return "Failed lol";
```

Comment:
Not sure this is the best failure code ... empty string instead?

Reply:
Not sure what you mean here. If the gchar* is empty and we attempt to create a c++ std::string, the program will crash. Therefore, we need to check if it is empty before we construct the string. I added a check to ensure type is not nullptr.

Reply:
Oh wait the failed_lol. True.

Reply:
oh oops forgot to get rid of this. Set it to empty string and added checks where it is called